CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,474 @@
+
+ODIS currently makes available its OPRF function with the following API:
+
+```typescript
+// POST to /getBlindedMessagePartialSig
+// Headers:
+//   Content-Type: application/json
+//   Authorization: ${signature from authorized signer (e.g. DEK) for account}
+interface GetBlindedMessageSigRequest {
+  // Celo account address. Query is charge against this account's quota.
+  account: string
+  // Query message. A blinded elliptic curve point encoded in base64.
+  blindedQueryPhoneNumber: string
+  // Optional on-chain identifier. Unlocks additional quota if the
+  // account is verified as an owner of the identifier.
+  hashedPhoneNumber?: string
+  // [Deprecated] Client-specified request timestamp, in milliseconds.
+  // Used for request expiration.
+  timestamp?: number
+  // Client-specified session ID.
+  sessionID?: string
+}
+```
+
+#### Extension for domain-restricted requests
+
+In order to support domain-restricted queries, a new API is proposed:
+
+```typescript
+// POST to /getDomainRestrictedPartialSig
+// Headers:
+//   Content-Type: application/json
+interface GetDomainRestrictedSigRequest {
+  // Domain specification.
+  // Selects the PRF domain and rate limiting rules.
+  domain: Domain
+  // Domain-specific options.
+  // Used for inputs relevant to the domain, but not part of the domain string.
+  // Example: { "authorization": <signature> } for an account-restricted domain.
+  domainOptions?: DomainOptions
+  // Query message. A blinded elliptic curve point encoded in base64.
+  blindedMessage: string
+  // Client-specified session ID.
+  sessionID?: string
+}
+
+interface Domain {
+  // Unique name of the domain. (e.g. "ODIS Password Domain")
+  name: string
+  // Major version number. Allows for backwards incompatible changes.
+  version: number
+  // Arbitrary key-value pairs.
+  // MUST be serializable to EIP-712 encoding.
+  [key: string]: EIP712Value
+}
+
+interface DomainOptions {
+  // Arbitrary key-value pairs.
+  // Valid keys and values depend on the domain specification.
+  // SHOULD be serializable to EIP-712 encoding.
+  [key: string]: EIP712Value
+}
+```
+
+### Domain specifiers
+
+Domain specifiers serve two functions:
+
+- Define the input domain, such that it is uncorrelated to the output in any other domain.
+- Specify the rate limiting function to be used in deciding whether to serve the request.
+
+In order to provide flexibility and support future extensions, the domain specifier is structured
+and allows for the usage of multiple domain-specific rate limiting rules, as support by the ODIS
+service.
+
+In the domain specifiers, two fields MUST be provided: `name` and `version`. Together, these select
+which set of rules the service will apply to the request. Rule sets are included in the ODIS
+implementation, and new rule sets may be added over time. Updating a rule set in a way that would
+change the rules applied to existing domains MUST be accompanied by an increase in the `version`
```

Comment:
Currently we have the freedom to tweak rate limits as needed. I.e. if we want to add cEUR balances to quota calculation, we can do that easily. I worry that this requirement may be too strict if it implies that we would need to bump the version number for the domain to make such a change, since bumping the version number changes the output. That is, users who queried ODIS for idenitifiers before we added support for cEUR would not be able to use the updated rate limit without switching to a new identifier. Maybe we can say that any change which _restricts_ a rate limit SHOULD be accompanied by an increase in the version?

Reply:
I'd actually apply that rule in the opposite direction. I'm not worried about changes that restrict the rate limit (i.e. such that previously valid requests will not longer be valid), I am worried about changes that loosen the rules. Making a previously invalid request valid means the user is exposed to new risk, which you can argue should only happen by opting in. I agree that its beneficial to have more flexibility though, using cEUR for the phone number domain seems fairly natural, and I'll change this to a SHOULD.